File: packages/gemi-kotlin/gemi-chat/src/main/kotlin/dev/gemijs/chat/Transport.kt
package dev.gemijs.chat

import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.job
import kotlinx.coroutines.suspendCancellableCoroutine
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response

/** A request `ChatSession` makes. Always a POST. */
public class ChatRequest(
  public val url: String,
  public val headers: Map<String, String>,
  public val body: ByteArray,
  public val contentType: String,
)

public class ChatResponse(public val statusCode: Int, public val body: Flow<ByteArray>) {
  public val isSuccess: Boolean get() = statusCode in 200..299

  /** The whole body, for the responses that are not streams. */
  public suspend fun bytes(): ByteArray {
    var all = ByteArray(0)
    body.collect { all += it }
    return all
  }
}

/**
 * What `ChatSession` sends requests through.
 *
 * `OkHttpTransport` is the one an app uses. The seam exists so a test can
 * stand in for the server, and so an app with its own networking stack can put
 * it underneath.
 */
public interface ChatTransport {
  /**
   * Sends the request. The body is streamed: the agent routes answer with SSE,
   * and a turn's frames have to reach the transcript as they arrive.
   *
   * Cancelling the calling coroutine — or the one collecting the body — must
   * close the connection. That is what `stop()` relies on to stop the UI at
   * once, before the server has even been told.
   */
  public suspend fun send(request: ChatRequest): ChatResponse
}

/** `ChatTransport` over OkHttp. */
public class OkHttpTransport(
  private val client: OkHttpClient =
    OkHttpClient.Builder()
      // Between two frames, not for the whole stream: a model thinking before
      // it speaks is silent for a while, and the server's keepalives are what
      // keep this from firing on a live run.
      .readTimeout(60, TimeUnit.SECONDS)
      .build()
) : ChatTransport {
  override suspend fun send(request: ChatRequest): ChatResponse {
    val call =
      client.newCall(
        Request.Builder()
          .url(request.url)
          .apply { request.headers.forEach { (key, value) -> header(key, value) } }
          .post(request.body.toRequestBody(request.contentType.toMediaType()))
          .build()
      )
    val response = call.await()
    val body =
      flow {
          // Closing the call is the only thing that interrupts a blocking read.
          currentCoroutineContext().job.invokeOnCompletion { call.cancel() }
          response.use {
            val source = it.body.source()
            val buffer = ByteArray(8192)
            while (true) {
              val read = source.inputStream().read(buffer)
              if (read == -1) break
              emit(buffer.copyOf(read))
            }
          }
        }
        .flowOn(Dispatchers.IO)
    return ChatResponse(response.code, body)
  }
}

private suspend fun Call.await(): Response = suspendCancellableCoroutine { continuation ->
  continuation.invokeOnCancellation { cancel() }
  enqueue(
    object : Callback {
      override fun onResponse(call: Call, response: Response) {
        continuation.resume(response) { _, value, _ -> value.close() }
      }

      override fun onFailure(call: Call, e: IOException) {
        continuation.resumeWithException(e)
      }
    }
  )
}

File: packages/gemi-swift/Sources/GemiChat/Transport.swift
import Foundation

/// What `ChatSession` sends requests through.
///
/// `URLSessionTransport` is the one an app uses. The seam exists so a test can
/// stand in for the server, and so an app with its own networking stack — a
/// certificate pin, a proxy, a request signer — can put it underneath.
public protocol ChatTransport: Sendable {
  /// Sends `request`. The body is streamed: the agent routes answer with SSE,
  /// and a turn's frames have to reach the transcript as they arrive.
  ///
  /// Cancelling the calling task must close the connection. That is what
  /// `stop()` relies on to stop the UI at once, before the server has even
  /// been told.
  func send(_ request: URLRequest) async throws -> ChatResponse
}

public struct ChatResponse: Sendable {
  public var statusCode: Int
  public var body: AsyncThrowingStream<Data, Error>

  public init(statusCode: Int, body: AsyncThrowingStream<Data, Error>) {
    self.statusCode = statusCode
    self.body = body
  }

  public var isSuccess: Bool { (200..<300).contains(statusCode) }

  /// The whole body, for the responses that are not streams.
  public func data() async throws -> Data {
    var data = Data()
    for try await chunk in body { data.append(chunk) }
    return data
  }
}

/// `ChatTransport` over `URLSession`.
public struct URLSessionTransport: ChatTransport {
  public var session: URLSession

  public init(session: URLSession = .shared) {
    self.session = session
  }

  public func send(_ request: URLRequest) async throws -> ChatResponse {
    let (bytes, response) = try await session.bytes(for: request)
    let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
    let body = AsyncThrowingStream<Data, Error> { continuation in
      let task = Task {
        var chunk = Data()
        chunk.reserveCapacity(4096)
        do {
          for try await byte in bytes {
            chunk.append(byte)
            // A frame only completes on a newline, so that is where a chunk is
            // worth handing on; the cap keeps a long line from buffering whole.
            if byte == 0x0A || chunk.count >= 4096 {
              continuation.yield(chunk)
              chunk.removeAll(keepingCapacity: true)
            }
          }
          if !chunk.isEmpty { continuation.yield(chunk) }
          continuation.finish()
        } catch {
          continuation.finish(throwing: error)
        }
      }
      continuation.onTermination = { _ in task.cancel() }
    }
    return ChatResponse(statusCode: statusCode, body: body)
  }
}
